tests/workload/transactions.go

```diff
@@ -0,0 +1,125 @@
+// Copyright (C) 2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package workload
+
+import (
+	"context"
+	"time"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/ava-labs/hypersdk/chain"
+	"github.com/ava-labs/hypersdk/rpc"
+	"github.com/ava-labs/hypersdk/utils"
+)
+
+// TxWorkloadFactory prescribes an exact interface for generating transactions to test on a given environment
+// and a sized sequence of transactions to test on a given environment and reach a particular state
+type TxWorkloadFactory interface {
+	// Generate a new TxWorkloadIterator that generates a single sequence of transactions
+	// and corresponding assertions.
+	NewBasicTxWorkload(uri string) (TxWorkloadIterator, error)
+	// Generates a new TxWorkloadIterator that generates a sequence of transactions of the given size.
+	NewSizedTxWorkload(uri string, size int) (TxWorkloadIterator, error)
+}
+
+// TxWorkloadIterator provides an interface for generating a sequence of transactions and corresponding assertions.
+// The caller must proceed in the following sequence:
+// 1. Next
+// 2. GenerateTxWithAssertion
+// 3. (Optional) execute the returned assertion on an arbitrary number of URIs
+type TxWorkloadIterator interface {
+	// Next returns true iff there are more transactions to generate.
+	Next() bool
+	// GenerateTxWithAssertion generates a new transaction and an assertion function that confirms
+	// 1. The tx was accepted on the provided URI
+	// 2. The state was updated as expected according to the provided URI
+	GenerateTxWithAssertion(context.Context) (*chain.Transaction, func(ctx context.Context, uri string) error, error)
+}
+
+func ExecuteWorkload(ctx context.Context, require *require.Assertions, uris []string, generator TxWorkloadIterator) {
+	submitClient := rpc.NewJSONRPCClient(uris[0])
+
+	for generator.Next() {
+		tx, confirm, err := generator.GenerateTxWithAssertion(ctx)
+		require.NoError(err)
+
+		_, err = submitClient.SubmitTx(ctx, tx.Bytes())
+		require.NoError(err)
+
+		utils.ForEach(func(uri string) {
+			err := confirm(ctx, uri)
+			require.NoError(err)
+		}, uris)
+	}
+}
+
+func GenerateNBlocks(ctx context.Context, require *require.Assertions, uris []string, factory TxWorkloadFactory, n uint64) {
+	uri := uris[0]
+	generator, err := factory.NewSizedTxWorkload(uri, int(n))
+	require.NoError(err)
+	client := rpc.NewJSONRPCClient(uri)
+
+	_, startHeight, _, err := client.Accepted(ctx)
+	require.NoError(err)
+	height := startHeight
+	targetheight := startHeight + n
+
+	for generator.Next() && height < targetheight {
+		tx, confirm, err := generator.GenerateTxWithAssertion(ctx)
+		require.NoError(err)
+
+		_, err = client.SubmitTx(ctx, tx.Bytes())
+		require.NoError(err)
+
+		require.NoError(confirm(ctx, uri))
+
+		_, acceptedHeight, _, err := client.Accepted(ctx)
+		require.NoError(err)
+		height = acceptedHeight
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, 10*time.Second)
```

Comment:
Maybe we can just expect the caller-provided context to specify the timeout?

Reply:
Deleting